monai/losses/image_dissimilarity.py

```diff
@@ -0,0 +1,153 @@
+        else:
+            raise ValueError(
+                f'Unsupported kernel_type: {kernel_type}, available options are ["rectangular", "triangular", "gaussian"].'
+            )
+
+    def make_rectangular_kernel(self):
+        shape = [1, self.in_channels] + [self.kernel_size] * self.ndim
+        return torch.ones(shape, dtype=torch.float), self.kernel_size ** self.ndim, int((self.kernel_size - 1) / 2)
+
+    def make_triangular_kernel(self):
+        fsize = torch.tensor((self.kernel_size + 1) / 2, dtype=torch.int)
+        f1 = torch.ones([1, 1] + [fsize] * self.ndim, dtype=torch.float) / fsize  # (1, 1, D, H, W)
+        f2 = (
+            torch.ones([self.in_channels, 1] + [fsize] * self.ndim, dtype=torch.float) / fsize
+        )  # (1, in_channels, D, H, W)
+        # (1, 1, D, H, W) -> (1, in_channels, D, H, W)
+        fn = conv_dict[self.ndim]
+        kernel = fn(f1, f2, padding=int((fsize - 1) / 2))
+
+        return kernel, torch.sum(kernel ** 2), int((fsize - 1) / 2)
+
+    def make_gaussian_kernel(self):
+        mean = (self.kernel_size - 1) / 2.0
+        sigma = self.kernel_size / 3
+
+        grid_ndim = torch.arange(0, self.kernel_size)
+        grid_ndim_ch = torch.arange(0, self.in_channels)
+
+        if self.ndim == 1:
+            grid = torch.meshgrid(grid_ndim_ch, grid_ndim)
+        elif self.ndim == 2:
+            grid = torch.meshgrid(grid_ndim_ch, grid_ndim, grid_ndim)
+        elif self.ndim == 3:
+            grid = torch.meshgrid(grid_ndim_ch, grid_ndim, grid_ndim, grid_ndim)
+        else:
+            raise ValueError
+
+        grid = torch.stack(grid, dim=-1).to(dtype=torch.float)
+        kernel = torch.exp(-torch.sum(torch.square(grid - mean), dim=-1) / (2 * sigma ** 2)).unsqueeze(
+            0
+        )  # (1, in_channel, kernel_size, kernel_size, kernel_size)
+        return kernel, torch.sum(kernel ** 2), int((self.kernel_size - 1) / 2)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            input: the shape should be BNH[WD].
+            target: the shape should be BNH[WD].
+        Raises:
+            ValueError: When ``self.reduction`` is not one of ["mean", "sum", "none"].
+        """
+        assert (
+            input.shape[1] == self.in_channels
+        ), f"expecting input with {self.in_channels} channels, got input of shape {input.shape}"
+        assert (
+            input.ndim - 2 == self.ndim
+        ), f"expecting input with {self.ndim} spatial dimensions, got input of shape {input.shape}"
+        assert (
+            target.shape == input.shape
+        ), f"ground truth has differing shape ({target.shape}) from input ({input.shape})"
+
+        t2, p2, tp = target ** 2, input ** 2, target * input
+
+        # sum over kernel
+        fn = conv_dict[self.ndim]
+        t_sum = fn(target, weight=self.kernel, padding=self.padding)
```

Comment:
perhaps separable filtering is faster in some cases, would be interesting to have a benchmarking? https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/networks/layers/simplelayers.py#L166-L168

Reply:
I will try that : )

Reply:
> perhaps separable filtering is faster in some cases, would be interesting to have a benchmarking?
> 
> https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/networks/layers/simplelayers.py#L166-L168

yup - pity that deepreg had separable gaussian kernel implemented in a file next to it ;) @acasamitjana @mathpluscode, a ticket?

Reply:
actually, we still have it https://github.com/DeepRegNet/DeepReg/blob/7c22a30ef27d7ff1bec5d21cbdb1a25f65cd5027/deepreg/model/loss/label.py#L314

Reply:
I think separable convolutions for the three kernels can be interesting. I have just opened a ticket on DeepReg and we can port it here

Reply:
separable filter implemented